[discounts] Added an option to change the bulk-site discounts model.

fajardoleo opened the pull request.

fajardoleo committed:
[discounts] Added an option to change the bulk-site discounts model.

swashata commented on src/components/FreemiusPricingMain.js:
```diff
@@ -713,6 +713,7 @@ class FreemiusPricingMain extends Component {
           billingCycles: Object.keys(billingCycles),
           currencies: Object.keys(currencies),
           currencySymbols: { usd: '$', eur: '€', gbp: '£' },
+          discountsModel: FSConfig.discounts_model,
```
@fajardoleo let's give this a default value in case this is not set in `FSConfig`, which keeps the current functionality.

```js
this.setState({
  // ...
  discountsModel: FSConfig.discounts_model ?? DiscountsModel.absolute,
  // ...
});
```

Reasons:

1. It is in general good practice to accept optional parameters on the top-level hierarchy instead of somewhere deep down. It is easier to debug and understand. 
2. With this change, you can now be strict at `calculateMultiSiteDiscount`, which would always expect the value of the parameter `discountsModel` to be set and not undefined.

I am sure, in time, we will see more values in the configuration enum like `alwaysRelatitveoMonthly`, `alwaysRelativeToAnnual` and having a default in the top-level would ease up the default implementation.

swashata commented on src/entities/Pricing.js:
```diff
@@ -21,6 +21,11 @@ export const BillingCycleString = {
   LIFETIME: 'lifetime',
 };
 
+export const DiscountsModel = {
```
Wonderful job @fajardoleo on thinking this ahead 👏 . A set contract with enums is definitely the way to go 😎

swashata commented on src/services/PlanManager.js:
```diff
@@ -1,4 +1,4 @@
-import { BillingCycle, Pricing } from '../entities/Pricing';
```
Since we are touching this file, let's do a refactoring. The way this `PlanManager` is made, isn't really ESModule way and it is very error prone.

I will explain the reason (I wrote everything down before and Grammarly just deleted all of the info) and will try to walk you through the process.

## Understanding ES Modules

Please read the documentations [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules#exporting_module_features). In general:

1. ESModules are scoped in the file they are written in. They are instantiated only once.
2. We can export functions, classes and variables from ESM files.


```js
export class Foo {};

export function fooFunc{};

export const fooVar = 'foo';

// or we can mix and match
class FooClass {}

const fooVar = 'foo';

function fooFunc() {}

export { FooClass, fooVar, fooFunc };
```

## Good Practices

In general ES Modules should be side-effect free. Let us consider the code (which is **BAD**) in lib.js

**lib.js**
```js
export let value = 'something';

export function setValue(newValue) {
  value = newValue;
}
```

Now we have another module

**module-loaded-later.js**
```js
import { value } from 'lib';

export function doSomething() {
  console.log(value);
}
```

Finally, in our main app, we do this:

**main.js**
```js
import { value, setValue } from 'lib';

console.log(value); // will output 'something'
setValue('notSomething');
console.log(value); // what would it output???????

import('module-loaded-later').then(({doSomething}) => {
 doSomething(); // how about now?
});
```

I leave it up to you to run and figure what's going to happen. I am sure you will catch on why having side-effects is a bad idea.

## Why this file has a bad side-effect

Let's say in **sub.js** we do this.

```js
import { getInstance } from 'PlanManager';

const PlanManager = getInstance(subPlans);

export function checkPlansPricings() {
  return PlanManager.foo().someThing();
}
```

Now in **main.js**, we have

```js
import { checkPlansPricings } from 'sub';
import { getInstance } from 'PlanManager';

// do something with sub
if (!checkPlansPricings()) {
  const newPlansManager = getInstance(newPlans);
}
```

What do you think would `newPlansManager` be bound to? `subPlans` or `newPlans`? Because of the code below

```js
function getInstance(plans) {
  if (null !== _instance) {
    return _instance;
  }
  // ...
}
```

It would still be bound to `subPlans`, as we did the side-effect right when `import`ing the module.

The only reason we didn't face this error right now, is because we never had multiple plans. But regardless, we now have an excellent opportunity to fix and refactor this.

## Proposal for Fixes

Of course, we could change the `getInstance` behavior to give a fresh instance bound to `plans`, but we can simplify it further and better. We can just export a class and instantiate in place. We already have all the private and public methods written. Instead of `_plans`, we would have to refactor to `this.plans`.

**PlanManager.js**
```js
export default class PlanManager {
  constructor(plans) {
    this.plans = plans;
      
    // maybe call some internals to set the data structure
    this.#setPricingSortedByLicensesAsc();
   }

  calculateMultiSiteDiscount() {...}

  #setPricingSortedByLicensesAsc() {...}
}
```

Then for usage

```js
import PlanManager from 'PlanManager';

// later in some component life-cycle
this.planManager = new PlanManager(this.plans);
```

Please keep in mind

- The file should be side-effect free, there shouldn't be any module level variables which would mutate.
- Please avoid straight passthrough functions like below which makes no sense.

<img width="719" alt="image" src="https://user-images.githubusercontent.com/1623381/165718109-d9dbfe3e-1dd9-48e5-9669-d32b67ef3a58.png">


I checked it is being used at only three places. (Right click on `PlanManager` > Peek > References`).

<img width="675" alt="image" src="https://user-images.githubusercontent.com/1623381/165717664-c2d6ca2f-16ea-4d46-9814-b6f1862595c2.png">

So it would be an easy fix and we will learn in the process.

swashata commented on src/services/PlanManager.js:
```diff
@@ -85,11 +85,17 @@ function getInstance(plans) {
         pricingBillingFrequency = BillingCycle.MONTHLY;
       }
 
+      const undiscountedPrice = (singleSitePrice * pricing.licenses);
```
Shouldn't the `undiscountedPrice` be calculated based on `DiscountsModel`? Here we are calculating for `RELATIVE` only, but later on we are diving by `ABSOLUTE` if asked. Are you sure this is the behavior we want?

swashata reviewed (changes requested):
@fajardoleo Please see the suggestions.

swashata committed:
[style] [ui] [fix] Fix some minor UI glitch + Improve the default style with new brand color.

swashata committed:
[trial] [paid] Redirect to the checkout page if trial requires subscription.

swashata committed:
[ajax-loader] [ui] [enhance] Some misc UI enhancements:

1. Use the same AJAX loader that comes with the SDK when embedded
   inside the Dashboard.
2. Fine tune some styling and border radius to make it consistent with
   the design.

swashata committed:
[trial] [non-opt-in] [loader] Fix issue for activating trial when user is not opted in + Fix DOM prop bug in the loader.

swashata committed:
[color-scheme] [wp-admin] Use native WP Admin color schemes.

swashata committed:
[theming] [update] Use consistent theming as the checkout.

swashata committed:
[cta] [fix] Fix logic for determining plan change and CTA labels.

swashata committed:
[ui] [update] Make design closer to WP admin.

swashata committed:
[cta] [label] Fix logical typo for determining the "Now" suffix.
